core/storage/btree.rs

```diff
@@ -1191,83 +1198,191 @@ impl BTreeCursor {
 
+        // Check if we're about to exceed usable space
+        // We also need some buffer space for future cell pointers
+        let required_space = used_space + self.estimate_cell_pointer_growth();
+        required_space > available_space
+    }
+
+    fn is_underflow(&self, page: &PageContent) -> bool {
+        // Root is special case - only underflow when empty with one child
+        let current_page = self.stack.top();
+        if current_page.get().id == self.root_page {
+            return page.cell_count() == 0 && !page.is_leaf();
+        }
+
+        let used_space = self.calculate_used_space(page);
+        let usable_space = self.usable_space();
+
+        // Underflow if used space is less than 50% of usable space
+        used_space < (usable_space / 2)
+    }
+
+    fn balance_root(&mut self) {
+        let current_root = self.stack.top();
+        let contents = current_root.get().contents.as_mut().unwrap();
+        let is_page_1 = current_root.get().id == 1;
         let offset = if is_page_1 { DATABASE_HEADER_SIZE } else { 0 };
-        let new_root_page = self.allocate_page(PageType::TableInterior, offset);
-        {
-            let current_root = self.stack.top();
-            let current_root_contents = current_root.get().contents.as_ref().unwrap();
 
-            let new_root_page_id = new_root_page.get().id;
-            let new_root_page_contents = new_root_page.get().contents.as_mut().unwrap();
-            if is_page_1 {
-                // Copy header
-                let current_root_buf = current_root_contents.as_ptr();
-                let new_root_buf = new_root_page_contents.as_ptr();
-                new_root_buf[0..DATABASE_HEADER_SIZE]
-                    .copy_from_slice(&current_root_buf[0..DATABASE_HEADER_SIZE]);
+        if self.is_underflow(contents) {
+            let child_page_id = contents.rightmost_pointer().unwrap();
+            let child_page = self
+                .pager
+                .read_page(child_page_id as usize)
+                .expect("This shouldn't have happened, child page not found");
+            let child_contents = child_page.get().contents.as_mut().unwrap();
+
+            // Defragment child before inserting its cells into root because root is smaller than child due to it
+            // containing header.
+            self.defragment_page(child_contents, self.database_header.borrow());
+
+            let necessary_space = self.calculate_used_space(child_contents);
+            let available_space = if is_page_1 {
+                self.usable_space() - DATABASE_HEADER_SIZE
+            } else {
+                self.usable_space()
+            };
+
+            // If root can't consume child, leave it empty temporarily and handle it in balance_leaf step
+            if available_space < necessary_space {
+                return;
+            }
+
+            let grandchild_ptr = child_contents.rightmost_pointer();
```

Comment:
rename `grandchild_ptr` -> `child_rightmost_pointer`? There could be a lot of grand childs so it this feels more exact.

Reply:
Moreover, if we are balancing root because it doesn't have any more cells, why do we expect the child to not be a leaf page? I mean, if you take rightmost pointer it means that child is a interior page, therefore, the tree is not balanced.